File: 3721-count-mentions-per-user/3721-count-mentions-per-user.erl
-spec create_user_list(N :: integer()) -> [tuple()].
create_user_list(N)     -> create_user_list(0, N).

-spec create_user_list(I :: integer(), N :: integer()) -> [tuple()].
create_user_list(N, N)  -> [];
create_user_list(I, N)  -> [{I, -1, 0}] ++ create_user_list(I + 1, N).


-spec set_user_online(Users :: [tuple()], CurrentTime :: integer()) -> [tuple()].
set_user_online(Users, CurrentTime) ->
    lists:map(
        fun (User) ->
            {Id, TTO, MentionCount} = User,

            if
                TTO =< CurrentTime -> {Id, -1, MentionCount};
                true -> User
            end
        end, Users
    ).


-spec set_user_offline(Users :: [tuple()], CurrentTime :: integer(), TargetId :: integer()) -> [tuple()].
set_user_offline(Users, CurrentTime, TargetId) ->
    lists:map(
        fun (User) ->
            case User of
                {TargetId, _, MentionCount} -> {TargetId, CurrentTime + 60, MentionCount};
                _ -> User
            end
        end, Users
    ).


-spec is_online(OTT :: integer()) -> integer().
is_online(-1)   -> 1;
is_online(_)    -> 0.


-spec split(Str :: string(), Delim :: char()) -> [string()].
split(Str, Delim)               -> split(Str, [], Delim).

-spec split(Str :: string(), Tmp :: string(), Delim :: char()) -> [string()].
split([], [], _)                -> [];
split([], Tmp, Delim)           -> [Tmp];
split([Delim | T], Tmp, Delim)  -> [Tmp] ++ split(T, [], Delim);
split([H | T], Tmp, Delim)      -> split(T, Tmp ++ [H], Delim).


-spec freq(List :: [integer()]) -> map().
freq(List)          -> freq(List, maps:new()).

-spec freq(List :: [integer()], Map :: map()) -> map().
freq([], Map)       -> Map;
freq([H|T], Map)    -> freq(T, maps:update_with(H, fun(X) -> X + 1 end, 1, Map)).


-spec make_inc_list_with_ids(N :: integer(), Ids :: [integer()]) -> [integer()].
make_inc_list_with_ids(N, Ids) -> 
    IdsMap = freq(Ids),
    NSIncList = maps:to_list(
        lists:foldl(
            fun(I, Map) ->
                case maps:is_key(I, Map) of
                    true -> Map;
                    false -> maps:put(I, 0, Map)
                end
            end, IdsMap, lists:seq(0, N - 1)
        )
    ),
    SIncList = lists:keysort(1, NSIncList),
    lists:map(fun(E) -> element(2, E) end, SIncList).


-spec increase_mention_count(User :: [tuple()], IncList :: [integer()]) -> [tuple()].
increase_mention_count(User, IncList) ->
    lists:map(
        fun ({{Id, TTO, MentionCount}, Inc}) ->
            {Id, TTO, MentionCount + Inc}
        end, lists:zip(User, IncList)
    ).


-spec process_message(Users :: [tuple()], Mentions :: string()) -> [tuple()].
process_message(Users, Mentions) ->
    N = length(Users),
    IncList = case hd(Mentions) of
        $A -> 
            lists:duplicate(N, 1);
        $H -> 
            lists:map(
                fun ({_, TTO, _}) ->
                    is_online(TTO)
                end, Users
            );
        _ -> 
            Ids = split(Mentions, 32), %32 is whitespace
            IntIds = lists:sort(
                lists:map(
                    fun (Id) ->
                        list_to_integer(element(2, lists:split(2, Id)))
                    end, Ids       
                )
            ),
            make_inc_list_with_ids(N, IntIds)
    end,
    increase_mention_count(Users, IncList).


-spec parse_event(Users :: [tuple()], Event :: [tuple()]) -> [tuple()].
parse_event(Users, Event) ->
    {Msg, CurrentTime, Mentions} = Event,

    OnlinedUsers = set_user_online(Users, CurrentTime),

    case hd(Msg) of
        $M ->
            process_message(OnlinedUsers, Mentions);
        $O ->
            Id = list_to_integer(Mentions),
            set_user_offline(OnlinedUsers, CurrentTime, Id)
    end.


-spec count_mentions(N :: integer(), Es :: [[unicode:unicode_binary()]]) -> [integer()].
count_mentions(N, Es) ->
    NewEs = lists:map(
        fun (Event) ->
            [Message | [TimeStamp | [Mentions | _]]] = lists:map(fun binary_to_list/1, Event),
            {Message, list_to_integer(TimeStamp), Mentions}
        end, Es
    ),

    SortedEs = lists:sort(
        fun(A, B) ->
            {AMsg, ATime, _} = A,
            {BMsg, BTime, _} = B,

            if
                ATime == BTime ->
                    if
                        hd(BMsg) == $O -> false;
                        true -> true
                    end;
                true -> ATime =< BTime
            end
        end, NewEs
    ),

    Us = create_user_list(N),

    NewUs = lists:foldl(
        fun(Event, Users) ->
            parse_event(Users, Event)
        end, Us, SortedEs
    ),

    lists:map(
        fun(User) -> 
            element(3, User)
        end, NewUs
    ).

File: 3721-count-mentions-per-user/3721-count-mentions-per-user.cpp
class Solution {
public:
    struct user{
        bool online = true;
        // will be online at this time
        int tto = -1;
        int mention_count = 0;
    };
    
    void set_user_online(vector<struct user>& users, int current_time){
        for(int i = 0; i < users.size(); ++i){
            // check tto and set user online which current time is after tto
            if(users[i].tto > current_time) continue;
            users[i].online = true;
            users[i].tto = -1;   
        }
    }
    
    void set_user_offline(vector<struct user>& users, int current_time, int id){
        users[id].online = false;
        users[id].tto = current_time + 60;
    }
    
    void process_message(vector<struct user>& users, const string& mention){
        switch(mention[0]){
            case 'A':
                for(int i = 0; i < users.size(); ++i) ++users[i].mention_count;
                break;
            case 'H':
                for(int i = 0; i < users.size(); ++i) if(users[i].online) ++users[i].mention_count;
                break;
            default:
                // 1. split and get user ids
                // 2. count online user in (1) list
                
                // split
                vector<int> ids{};
                stringstream ss{};
                
                for(int i = 0; i < mention.length(); ++i){
                    if(mention[i] == ' '){
                        ids.push_back(stoi(ss.str().substr(2)));
                        ss.str("");
                        continue;
                    }
                    
                    ss << mention[i];
                }
                
                ids.push_back(stoi(ss.str().substr(2)));
                ///////////
                
                for(int i = 0; i < ids.size(); ++i) ++users[ids[i]].mention_count;
                break;
        }
    }
    
    void parse_event(vector<struct user>& users, const vector<string>& e){
        // 0 : MESSAGE or OFFLINE
        // 1 : TIME STAMP
        // 2 : MENTION ( id<> or HERE or ALL)
        //  id<N>: mention specific user
        //  HERE: mention only online user
        //  ALL: mention users whether online or not
        
        // make current time
        int current_time = stoi(e[1]);
        
        set_user_online(users, current_time);
        
        switch(e[0][0]){
            case 'M':
                // process message
                process_message(users, e[2]);
                break;
            case 'O':
                // process offline
                int id = stoi(e[2]);
                set_user_offline(users, current_time, id);
                break;
        }
    }
    
    vector<int> countMentions(int n, vector<vector<string>>& e) {
        vector<struct user> users;
        users.resize(n);
        
        sort(e.begin(), e.end(), [](vector<string>& a, vector<string>& b)->bool{
            int a_time = stoi(a[1]), b_time = stoi(b[1]);
            
            if(a_time == b_time){
                if(a[0][0] == 'O') return true;
                if(b[0][0] == 'O') return false;
            }
            
            return a_time <= b_time;
        });
        
        for(int i = 0; i < e.size(); ++i){
            parse_event(users, e[i]);
        }
        
        vector<int> res{};
        for(int i = 0; i < n; ++i) 
            res.emplace_back(users[i].mention_count);
        
        return res;
    }
};